### PR TITLE
Add support for separators in text copied from `SelectableRegion`

### PR DESCRIPTION
**Pull request by Rexios80**
Right now, copying text in a `SelectionArea` copies it with no separation. Since `SelectionArea` is very useful with vertical lists of `Text` widgets this is not ideal. This PR adds a `separator` field to `SelectionArea` so that developers can add separators such as newlines between selected `Text` widgets.

https://github.com/flutter/flutter/issues/104548

I added a new test for this

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by Rexios80** [comment]
It wasn't sitting well with me that nested `SelectableRegion`s didn't coordinate selection with parent `SelectableRegion`s. 

With composition it is now possible to have the following widget tree

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    const MaterialApp(
      home: Scaffold(
        body: SelectionArea(
          separator: '\n',
          child: Column(
            children: <Widget>[
              Text('How are you?'),
              SelectionArea(
                separator: ' ',
                child: Row(
                  children: <Widget>[
                    Text('Good, and you?'),
                    Text('Fine, thank you.'),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    ),
  );
}

```

copy/paste this

```
How are you?
Good, and you? Fine, thank you.
```

**Review by chunhtai — COMMENTED** [review]
This API will hardcode the behavior for entire selection area which is not something we want to do. I think the separator should be added into MultiSelectableSelectionContainerDelegate and then each multichild widget such as Row or column should wrap its own selection container and set the separate accordingly.

Another thing to consider is whether we should convert to copy html text instead, some desktop platforms can store html into clipboard, but i have not tested them all.

**Comment by Rexios80** [comment]
This behavior should definitely be developer configurable. I don't think we should do anything web-specific here such as copying html. I'll look into doing something with rows/columns.

**Comment by Rexios80** [comment]
Actually, I think these changes already handle that issue. You can now nest `SelectionArea`s to override the parent's separator.

**Comment by chunhtai** [comment]
> This behavior should definitely be developer configurable. I don't think we should do anything web-specific here such as copying html. I'll look into doing something with rows/columns.

I have not looked into this in more detail but I think you can store html into clipboard in macos https://developer.apple.com/documentation/appkit/nspasteboardtypehtml?language=objc

This will be useful to store rich text and be able to reuse in other application that can handle pasting html

**Comment by chunhtai** [comment]
> If we did go the route of MultiChild widgets handling the separation, we would probably need to create `SelectableFlex`, `SelectabelRow`, `SelectableColumn`, etc in which case `SelectionArea` would have no reason to exist.

No, The Flex can wrap the subtree with a selection container similar to how scrollable does that. It conditionally wrap the selection container only if there is another selection container above.

**Comment by chunhtai** [comment]
> Actually, I think these changes already handle that issue. You can now nest `SelectionArea`s to override the parent's separator.
> 
> I'm not sure touching anything else makes sense.

I think you meant SelectionContainer, I don't think we have good support for nesting SelectionArea.

**Comment by Rexios80** [comment]
I made nesting `SelectionArea`s work in this PR, but I like what you're saying about making `Flex` handle this. I'll work on implementing that.

**Comment by Rexios80** [comment]
An issue I'm seeing with that approach is `Scrollable` is a `StatelessWidget` whereas `Flex` is a `MultiChildRenderObjectWidget` which doesn't have a build method to do the `SelectionContainer` logic in

**Comment by Rexios80** [comment]
What exactly is the issue if this code works?

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    const MaterialApp(
      home: Scaffold(
        body: SelectionArea(
          separator: '\n',
          child: Column(
            children: <Widget>[
              Text('How are you?'),
              SelectionArea(
                separator: ' ',
                child: Row(
                  children: <Widget>[
                    Text('Good, and you?'),
                    Text('Fine, thank you.'),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    ),
  );
}
```

**Comment by chunhtai** [comment]
> An issue I'm seeing with that approach is `Scrollable` is a `StatefulWidget` whereas `Flex` is a `MultiChildRenderObjectWidget` which doesn't have a `build` method to do the `SelectionContainer` logic in

You can make column and row a statelessWidget and in the build method wrapps SelectionContainer.

> What exactly is the issue if this code works?

there are two issues:

1. SelectionArea is basically a selectionContainer with gesture and selectionoverlay logic. but in this case, you don't need the later two, so the nested SelectionArea should be replaced with SelectionContainer

2. We don't expect developer to need to manually insert these nested SelectionArea themselves, for obvious cases like Column,Row, or vertical scrollable, it should hardcode the separator directly.

**Comment by Rexios80** [comment]
Thank you for explaining. This is my first PR on the flutter framework itself so I'm very new to understanding how all these pieces fit together. I wasn't sure if changing the inheritance of `Row`, `Column`, etc would be considered a breaking change.

**Comment by chunhtai** [comment]
I don't think so, as long as the customer tests do not fail, we don't consider it a breaking change. Also I don't think we promise the type of a widget to stay the same as long as it is a still a `Widget`

**Comment by Rexios80** [comment]
@chunhtai I put something together. Let me know your thoughts. I still need to do an implementation for `Stack`.

**Comment by Rexios80** [comment]
Since `Column` widgets now default to using `\n` as their separator, I had to change the existing test for `SelectableRegion`. The previous behavior was highly undesirable though, does this count as a breaking change?

**Comment by chunhtai** [comment]
> The previous behavior was highly undesirable though, does this count as a breaking change?

No our own test doesn't justify being a breaking change, and in this case, the test is testing a wrong behavior, so no worry.

On the other hand, can we do one thing at a time? I suggest we leave the stack out from this PR, and do it in the next

**Comment by chunhtai** [comment]
Also please re-request review using the github UI when this PR is ready to review.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 1]
```diff
@@ -4541,17 +4446,59 @@ class Flex extends MultiChildRenderObjectWidget {
...
+  TextBaseline? get textBaseline;
 
   /// {@macro flutter.material.Material.clipBehavior}
   ///
   /// Defaults to [Clip.none].
+  Clip get clipBehavior;
+}
+
+/// The render object used by [Flex].
+class FlexRenderObject extends MultiChildRenderObjectWidget with FlexMixin {
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/basic.dart:4464]: I don't think we need to support separator if developer directly use Flex. If people does require this feature, they can just use the SelectableRegionContainerDelegate

File: packages/flutter/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -1830,33 +1830,8 @@ abstract class SingleChildRenderObjectWidget extends RenderObjectWidget {
...
-///    its children in named slots.
-abstract class MultiChildRenderObjectWidget extends RenderObjectWidget {
-  /// Initializes fields for subclasses.
-  ///
-  /// The [children] argument must not be null and must not contain any null
-  /// objects.
-  const MultiChildRenderObjectWidget({ super.key, this.children = const <Widget>[] });
-
+/// Mixin to allow sharing the required properties of [MultiChildRenderObjectWidget]
+mixin MultiChildRenderObjectWidgetMixin {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/framework.dart:1834]: Is this change needed?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 3]
```diff
@@ -4815,6 +4921,7 @@ class Row extends Flex {
   /// must not be null.
   const Row({
     super.key,
+    super.selectionSeparator,
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/basic.dart:4924]: I don't think we need to expose this for now. I can't think of a good use case where they would want to change this. We can open this up if we have real use case

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 4]
```diff
@@ -1416,6 +1424,22 @@ class _SelectableRegionContainerDelegate extends MultiSelectableSelectionContain
   }
 }
 
+/// A mixin to standardize the usage of [selectionSeparator].
+mixin MultiSelectableSeparator {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/selectable_region.dart:1428]: should this built in to the MultiSelectableSelectionContainerDelegate instead? does not seem necessary to mix this onto widget

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 5]
```diff
@@ -1260,7 +1260,15 @@ class _DirectionallyExtendCaretSelectionAction<T extends DirectionalCaretMovemen
   }
 }
 
-class _SelectableRegionContainerDelegate extends MultiSelectableSelectionContainerDelegate {
+/// A delegate for the [SelectionContainer]s of [SelectableRegion]s and their
+/// children.
+class SelectableRegionContainerDelegate extends MultiSelectableSelectionContainerDelegate {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/selectable_region.dart:1265]: We need a more generic name if we open this up. Maybe call StaticMultiSelectableSelectionContainerDelegate.

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 6]
```diff
@@ -1260,7 +1260,15 @@ class _DirectionallyExtendCaretSelectionAction<T extends DirectionalCaretMovemen
   }
 }
 
-class _SelectableRegionContainerDelegate extends MultiSelectableSelectionContainerDelegate {
+/// A delegate for the [SelectionContainer]s of [SelectableRegion]s and their
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/selectable_region.dart:1263]: This is no longer true. The doc should mention this delegate is used when selectables in the subtree are fixed in screen position and don't scroll.

File: packages/flutter/lib/src/widgets/framework.dart (hunk 2 above)
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/framework.dart:1834]: Not necessarily, but again I wanted to share these properties with classes that no longer extend this

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 4 above)
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/selectable_region.dart:1428]: I made this a mixin to share it with other classes. If we aren't exposing this property on anything else we can remove it, but we do probably need it for the Stack implementation

**Comment by Rexios80** [comment]
What about Scrollable? Should we not expose the selection separator there either?

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 4 above)
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/selectable_region.dart]: I went ahead and removed this because of the Flex changes

File: packages/flutter/lib/src/widgets/basic.dart (hunk 3 above)
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/basic.dart]: Removed

File: packages/flutter/lib/src/widgets/basic.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/basic.dart]: Refactored

File: packages/flutter/lib/src/widgets/basic.dart [hunk 7]
```diff
@@ -4360,119 +4362,22 @@ class PositionedDirectional extends StatelessWidget {
...
-    this.textDirection,
-    this.verticalDirection = VerticalDirection.down,
-    this.textBaseline, // NO DEFAULT: we don't know what the text's baseline should be
-    this.clipBehavior = Clip.none,
-    super.children,
-  }) : assert(!identical(crossAxisAlignment, CrossAxisAlignment.baseline) || textBaseline != null, 'textBaseline is required if you specify the crossAxisAlignment with CrossAxisAlignment.baseline');
-  // Cannot use == in the assert above instead of identical because of https://github.com/dart-lang/language/issues/1811.
-
+/// Mixin to allow sharing the required properties of [FlexRenderObject]
+mixin FlexMixin {
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/basic.dart:4366]: Then I guess you can have a _FlexBase, then this should probably call AbstractFlex and should be an abstract class

File: packages/flutter/lib/src/widgets/basic.dart [hunk 8]
```diff
@@ -4994,7 +5075,7 @@ class Row extends Flex {
 ///    use a [Column] inside a scrolling container.
 ///  * [Spacer], a widget that takes up space proportional to its flex value.
 ///  * The [catalog of layout widgets](https://flutter.dev/widgets/layout/).
-class Column extends Flex {
+class Column extends StatelessWidget with MultiChildRenderObjectWidgetMixin, FlexMixin {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/basic.dart:5078]: Why MultiChildRenderObjectWidgetMixin is needed?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 9]
```diff
@@ -4799,7 +4827,7 @@ class Flex extends MultiChildRenderObjectWidget {
 ///    that may by sized smaller (leaving some remaining room unused).
 ///  * [Spacer], a widget that takes up space proportional to its flex value.
 ///  * The [catalog of layout widgets](https://flutter.dev/widgets/layout/).
-class Row extends Flex {
+class Row extends StatelessWidget with MultiChildRenderObjectWidgetMixin, FlexMixin {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/basic.dart:4830]: Why MultiChildRenderObjectWidgetMixin is needed?

File: packages/flutter/lib/src/widgets/framework.dart (hunk 2 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/framework.dart:1834]: If the goal is to get the children getter, I think put it in AbstractFlex and use 
{@template ..} and {@endtemplate} to reuse documentation would be fine

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 10]
```diff
@@ -100,7 +100,7 @@ class Scrollable extends StatefulWidget {
     this.restorationId,
     this.scrollBehavior,
     this.clipBehavior = Clip.hardEdge,
-  }) : assert(semanticChildCount == null || semanticChildCount >= 0);
+  })  : assert(semanticChildCount == null || semanticChildCount >= 0);
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/scrollable.dart:103]: revert this?

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 11]
```diff
@@ -274,6 +274,15 @@ class Scrollable extends StatefulWidget {
   /// Determined by the [axisDirection].
   Axis get axis => axisDirectionToAxis(axisDirection);
 
+  String get _selectionSeparator {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/scrollable.dart:277]: This should probably be set in _ScrollableSelectionContainerDelegate instead

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 12]
```diff
@@ -1425,6 +1432,18 @@ class _SelectableRegionContainerDelegate extends MultiSelectableSelectionContain
...
+  ///
+  /// The [selectionSeparator] is used to separate the text of each selectable in
+  /// [selectables].
+  MultiSelectableSelectionContainerDelegate({
+     this.selectionSeparator = '',
+  });
+
+  /// The separator used to separate the text of each selectable in
+  /// [selectables].
+  final String selectionSeparator;
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/selectable_region.dart:1445]: probably use a getter instead of final property will make it easier to apply to _ScrollableSelectionContainerDelegate

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 13]
```diff
@@ -1425,6 +1432,18 @@ class _SelectableRegionContainerDelegate extends MultiSelectableSelectionContain
 /// This class optimize the selection update by keeping track of the
 /// [Selectable]s that currently contain the selection edges.
 abstract class MultiSelectableSelectionContainerDelegate extends SelectionContainerDelegate with ChangeNotifier {
+  /// Creates a [MultiSelectableSelectionContainerDelegate].
+  ///
+  /// The [selectionSeparator] is used to separate the text of each selectable in
+  /// [selectables].
+  MultiSelectableSelectionContainerDelegate({
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/selectable_region.dart:1439]: If selectionSeparator then this is not needed

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 12 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/selectable_region.dart:1445]: also probably name this to selectedContentSepatator

File: packages/flutter/test/widgets/selectable_region_test.dart [hunk 14]
```diff
@@ -696,7 +732,8 @@ void main() {
           home: SelectableRegion(
             focusNode: focusNode,
             selectionControls: materialTextSelectionControls,
-            child: const Column(
+            child: const Flex(
```
Comment (chunhtai) [inline packages/flutter/test/widgets/selectable_region_test.dart:735]: Instead of changing this, I will probably update the expected copied text

Comment (Rexios80) [reply packages/flutter/test/widgets/selectable_region_test.dart:735]: I think the issue here is the new Row/Column are interfering with the selection detection in these tests

File: packages/flutter/lib/src/widgets/basic.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/basic.dart:4366]: Can you elaborate more on the class signatures you want _FlexBase and AbstractFlex to have?

Why do we need both _FlexBase and AbstractFlex?

File: packages/flutter/lib/src/widgets/scrollable.dart (hunk 11 above)
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/scrollable.dart:277]: Not exactly sure what you want here. We need access to the axis to decide what separator to use, and _ScrollableSelectionContainerDelegate doesn't have access to that.

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 13 above)
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/selectable_region.dart:1439]: Can you please elaborate on what you want done here?

File: packages/flutter/lib/src/widgets/basic.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/basic.dart:4366]: I just made a _FlexBase for now. Let me know if you want any changes

File: packages/flutter/lib/src/widgets/basic.dart (hunk 8 above)
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/basic.dart:5078]: Removed

File: packages/flutter/lib/src/widgets/basic.dart (hunk 9 above)
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/basic.dart:4830]: Removed

File: packages/flutter/lib/src/widgets/scrollable.dart (hunk 10 above)
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/scrollable.dart:103]: Done

File: packages/flutter/lib/src/widgets/scrollable.dart (hunk 11 above)
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/scrollable.dart:277]: I moved this logic to where we create the _ScrollableSelectionHandler. Let me know if you want any changes.

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 12 above)
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/selectable_region.dart:1445]: I need clarification on the other points before I can do anything with this

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 15]
```diff
@@ -1260,7 +1260,14 @@ class _DirectionallyExtendCaretSelectionAction<T extends DirectionalCaretMovemen
   }
 }
 
-class _SelectableRegionContainerDelegate extends MultiSelectableSelectionContainerDelegate {
+/// A delegate for selectables within non-scrolling widgets
+class StaticMultiSelectableSelectionContainerDelegate extends MultiSelectableSelectionContainerDelegate {
+  /// Create a [StaticMultiSelectableSelectionContainerDelegate]
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/selectable_region.dart:1265]:
```suggestion
  /// Creates a [StaticMultiSelectableSelectionContainerDelegate]
```

File: packages/flutter/test/widgets/selectable_region_test.dart (hunk 14 above)
Comment (chunhtai) [reply packages/flutter/test/widgets/selectable_region_test.dart:735]: No entirely sure what do you mean? The finder below doesn't seem to care about column/row

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 13 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/selectable_region.dart:1439]: I don't remember my previous comment as well, the code looks fine now

**Review by chunhtai — COMMENTED** [review]
Code LGTM, left a question on the test.

File: packages/flutter/test/widgets/selectable_region_test.dart (hunk 14 above)
Comment (Rexios80) [reply packages/flutter/test/widgets/selectable_region_test.dart:735]: Not at my computer right now to give actual logs, but if you run the tests without these changes you should see the failure.

Comment (Rexios80) [reply packages/flutter/test/widgets/selectable_region_test.dart:733]: Right, it throws range errors because the selections aren't found

```console
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following RangeError was thrown running a test:
RangeError (index): Invalid value: Valid value range is empty: 0

When the exception was thrown, this was the stack:
#0      List.[] (dart:core-patch/growable_array.dart:264:36)
#1      main.<anonymous closure>.<anonymous closure> (file:///Users/rexios/repos/flutter/packages/flutter/test/widgets/selectable_region_test.dart:1187:35)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

The test description was:
  can copy from toolbar
```

Comment (Rexios80) [reply packages/flutter/test/widgets/selectable_region_test.dart:733]: It might just be that the detection logic needs fixed, but I don't know enough about it to do it easily

Comment (Rexios80) [reply packages/flutter/test/widgets/selectable_region_test.dart]: Thanks! All the tests should pass now.

**Comment by Rexios80** [comment]
@chunhtai I don't have the button to re-request review, but I fixed the tests

File: packages/flutter/test/widgets/selectable_region_test.dart [hunk 16]
```diff
@@ -704,6 +740,7 @@ void main() {
           ),
         ),
       );
+      await tester.pump();
```
Comment (chunhtai) [inline packages/flutter/test/widgets/selectable_region_test.dart:743]: Can you add a comment on why the pump is needed?
i.e. The Column's SelectionContainer takes one frame to register to SelectionArea.

**Review by chunhtai — APPROVED** [review]
LGTM

**Review by chunhtai — APPROVED** [review]
still LGTM

**Comment by chunhtai** [comment]
cc Justin for secondary review

File: packages/flutter/lib/src/widgets/selectable_region.dart [hunk 17]
```diff
@@ -1849,12 +1868,8 @@ abstract class MultiSelectableSelectionContainerDelegate extends SelectionContai
     if (selections.isEmpty) {
       return null;
     }
-    final StringBuffer buffer = StringBuffer();
-    for (final SelectedContent selection in selections) {
-      buffer.write(selection.plainText);
-    }
     return SelectedContent(
-      plainText: buffer.toString(),
+      plainText: selections.map((SelectedContent e) => e.plainText).join(selectedContentSeparator),
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/selectable_region.dart:1872]: Just as a thought, what if instead of passing in a separator String, it was a function:

```
String Function(List<SelectedContent> selections)
```

Maybe this is overkill, but I wanted to talk through it.

**Review by justinmc — COMMENTED** [review]
Thanks for the PR, I definitely think you're right that we should support this.  I have two questions that I wanted to talk through, but otherwise I'm on board with this.

  1. Did you originally add selectedContentSeparator as a parameter to things like SelectionArea and then remove it?  Is there a conversation somewhere you can link me to about why it's not public like that?
  2. We could make this more flexible by passing a function instead of a String separator (see my comment below).  I'm not sure how many users would actually want it, though, so maybe it's not worth the added complexity.

File: packages/flutter/lib/src/widgets/selectable_region.dart (hunk 17 above)
Comment (Rexios80) [reply packages/flutter/lib/src/widgets/selectable_region.dart:1872]: I was also thinking that having a way for the separator to not be the same for the entire container might be a good idea

**Comment by Rexios80** [comment]
As for your second point, there is a conversation here: https://github.com/flutter/flutter/pull/121112#discussion_r1130195639

**Comment by chunhtai** [comment]
It looks like the ci is failing, can you fix it?

**Comment by Rexios80** [comment]
Those tests fail for me locally on the main branch

**Comment by Rexios80** [comment]
I deleted the whole flutter repo folder and cloned from the official repository. The tests still fail.

**Comment by chunhtai** [comment]
This looks like a valid failure
/packages/flutter/test/material/debug_test.dart: debugCheckHasScaffoldMessenger control test                                                                      
01:44 +1969 ~1: /b/s/w/ir/x/w/flutter/packages/flutter/test/material/debug_test.dart: debugCheckHasScaffoldMessenger control test 

you need to add an additional `Flex` to the test

**Comment by Rexios80** [comment]
@chunhtai All checks pass now except "Google testing"

**Comment by justinmc** [comment]
There is a legitimate failure there, but I'm trying to determine whether it's something we can work around in this PR or not.

**Comment by lrorpilla** [comment]
In regards to this feature, I would really like to be able to programmatically select certain portions in a `SelectableRegion`, given a start or end index of some text, and so on, as well as getting indices information when obtaining a `SelectableContent`.

I'm currently using a [**modified version**](https://github.com/flutter/flutter/pull/69461/files) of `SelectableText` which allows for clearing or setting the text selection and I really want a version of it that works for `SelectableArea`...

**Comment by Rexios80** [comment]
You can ask questions like this in any of the community groups listed here: https://flutter.dev/community

This definitely isn't the place

**Comment by Rexios80** [comment]
Well since we all just got notified anyways...

@justinmc are there any updates on the Google Testing issues?

**Comment by chunhtai** [comment]
I got distract by high prioirty issues. I am still investigating the issue in internal tests.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 18]
```diff
@@ -4934,16 +4965,69 @@ class Row extends Flex {
...
+      clipBehavior: clipBehavior,
+      children: children,
+    );
+
+    // Selection is only enabled when there is a parent registrar.
+    final SelectionRegistrar? registrar = SelectionContainer.maybeOf(context);
+    if (registrar != null) {
+      result = SelectionContainer(
+        registrar: registrar,
+        delegate: StaticMultiSelectableSelectionContainerDelegate(selectedContentSeparator: ' '),
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/basic.dart:5024]: I think StaticMultiSelectableSelectionContainerDelegate will need to be kept as state. otherwise, the delegate will gets recreated and the state inside it would be needed to be transfer again and again. Can you create another wrapper statefulwidget that hold the StaticMultiSelectableSelectionContainerDelegate as its state?

**Review by chunhtai — COMMENTED** [review]
There are at least three issues

1. _SelectionContainerState may concurrently modify the _listeners when calling them. This is a bug in SelectionRegistrant but got surfaced due to Column repeatedly creates the `StaticMultiSelectableSelectionContainerDelegate`. I will create a pr to fix this, but please also update the Column to not create the delegate every time a build is called
2. This change breaks tests in https://github.com/flutter/packages/tree/main/packages/flutter_markdown because the test expect Column not not create Flex, Can you check if there is anyway to migrate the tests?
3. Some dart syntax when omitting local variable type is enabled, I am not sure if this is a dart sdk issue or not, i have not looked at them all, but I will keep looking

**Comment by Rexios80** [comment]
> This change breaks tests in https://github.com/flutter/packages/tree/main/packages/flutter_markdown because the test expect Column not not create Flex, Can you check if there is anyway to migrate the tests?

How can I migrate those tests without this being merged in?

**Comment by Rexios80** [comment]
@chunhtai https://github.com/flutter/packages/pull/3693

**Comment by Rexios80** [comment]
Although it looks like Google testing is passing now???

**Comment by Hixie** [comment]
Have we run benchmarks on this? Making Row and Column into stateful widgets seems like a _really_ expensive change that will have huge impact across the board on Flutter apps. I would expect this to increase build times and increase memory usage. These are really basic widgets used all _over_ the place.

**Comment by Rexios80** [comment]
@Hixie Yeah I wasn't too sure about making `Row` and `Column` more complicated either. That's why the original changes had the separator handling on the `SelectionArea` itself.

**Review by chunhtai — COMMENTED** [review]
I don't think we should directly make Column into StatefulWidget, You can create a wrapper widget maybe call _SelectedTextSeparator, which is a statefulwidget. the Column should remain a stateless widget and only wrap the _FlexBase with _SelectedTextSeparator if there is a selectionContainer

**Comment by Hixie** [comment]
I'm very skeptical about any changes to Row or Column at all here. These are really core widgets that should not need to know about text selection at all, IMHO. They're just ways to configure the RenderFlex render object.

**Comment by chunhtai** [comment]
I agree with column should be simple, but that means people would have to do something extra to make column work correctly when selected.  think people would likely ignored it and make app with bad user experience.

Another way to solve this is to make the selection container smart enough to know whether to add a `\n` or not by looking at the comparing the rect for each selectable text. for example if the y axis  does not overlap, we add \n, if the x axis has enough gap we add \t.
This will be very opinionated and hard to customize though.

**Comment by Hixie** [comment]
That might be more reliable than having to teach every render object (or widget that wraps a render object) how to handle selection though. For example, how would we make it work for `Stack`? Surely that would have the same problem as `Flex`.

**Comment by Rexios80** [comment]
Should we just go back to what I had before where the `SelectionArea` handled it?

**Comment by chunhtai** [comment]
@Rexios80

I assume you meant adding an API in MultiSelectableSelectionContainerDelegate to customize how separator are inserted?

I think this is what we can do is that in MultiSelectableSelectionContainerDelegate.getSelectedContent when we concatenate SelectedContent, we should compare the coordinate of each selectable and add '\n' or '\t'. I think this should be enough for fixing most of the issue.

If people really want to change the separator their way, we can expose an API in MultiSelectableSelectionContainerDelegate that takes a list of Selectables and return a SelectedContent. This API will be called in getSelectedContent to get the result

**Comment by Rexios80** [comment]
Comparing coordinates seems fragile and like it might have a bunch of edge cases. What I meant is just doing what the original PR did which let you do this:

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    const MaterialApp(
      home: Scaffold(
        body: SelectionArea(
          separator: '\n',
          child: Column(
            children: <Widget>[
              Text('How are you?'),
              SelectionArea(
                separator: ' ',
                child: Row(
                  children: <Widget>[
                    Text('Good, and you?'),
                    Text('Fine, thank you.'),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    ),
  );
}
```

**Comment by Rexios80** [comment]
Users already have to wrap the widgets in `SelectionArea` for them to be selectable, so it's not a huge ask to just have them provide a separator if they want one. Adding separators automatically would be ideal, but currently there is no way to add separators at all and `SelectionArea` is mostly useless. Also if the field is on `SelectionArea` it is very easy for users to configure it.

**Comment by Rexios80** [comment]
Do we even have access to the information we need to make the separator logic automatic in `SelectableRegion`? All we have access to is the `Selectable`s which only have a way to read the selected content and nothing about the widget containing it.

**Comment by chunhtai** [comment]
There is a size https://api.flutter.dev/flutter/rendering/Selectable/size.html
and getTransformFrom
https://api.flutter.dev/flutter/widgets/SelectionContainerDelegate/getTransformFrom.html
this should be enough to calculate the rect of child Selectable in to the coordinate of SelectionContainerDelegate

**Comment by Rexios80** [comment]
@dsyrstad What about everything up to [bece955](https://github.com/flutter/flutter/pull/121112/commits/bece9551b235253d69d4c35875706d95a41f954d)?

That would be these changes: https://github.com/flutter/flutter/pull/121112/files/bece9551b235253d69d4c35875706d95a41f954d

**Comment by dsyrstad** [comment]
> @dsyrstad What about everything up to [bece955](https://github.com/flutter/flutter/pull/121112/commits/bece9551b235253d69d4c35875706d95a41f954d)?
> 
> That would be these changes: https://github.com/flutter/flutter/pull/121112/files/bece9551b235253d69d4c35875706d95a41f954d

@Rexios80  I'm not sure what the `registrar` change does and how it relates to implementing a separator. Can you explain?

**Comment by Rexios80** [comment]
That makes nested `SelectionArea` widgets work so you can have different separators for different portions of the text

**Comment by Rexios80** [comment]
Closing this PR in favor of https://github.com/flutter/flutter/pull/126835
